plat/common/lcpu.c

```diff
@@ -48,3 +191,447 @@ void ukplat_lcpu_halt_to(__nsec until)
+	return 1;
+}
+
+/* We want these to be externally defined as const to clarify that the vectors
+ * cannot be changed after initialization. However, we still need them non-const
+ * so we can still set them here. While we can do a DECONST and force allocation
+ * in .bss, we enter undefined behavior territory. So we just export a const
+ * pointer as proxy. This is still faster than calling a getter function and
+ * with LTO this will be optimized to a direct access.
+ */
+static unsigned long _lcpu_run_irqv;
+static unsigned long _lcpu_wakeup_irqv;
+
+const unsigned long * const lcpu_run_irqv = &_lcpu_run_irqv;
+const unsigned long * const lcpu_wakeup_irqv = &_lcpu_wakeup_irqv;
+
+int lcpu_mp_init(unsigned long run_irq, unsigned long wakeup_irq, void *arg)
+{
+	int rc;
+
+	/* Make sure this is run on the BSP only */
+	UK_ASSERT(lcpu_count == 1);
+	UK_ASSERT(lcpu_current_is_bsp());
+
+	/* Initialize architecture-dependent functionality. This will also do
+	 * CPU discovery and allocation
+	 */
+	rc = lcpu_arch_mp_init(arg);
+	if (unlikely(rc))
+		return rc;
+
+	/* Register the lcpu_run and lcpu_wakeup interrupt handlers */
+	rc = ukplat_irq_register(run_irq, lcpu_ipi_run_handler, NULL);
+	if (unlikely(rc)) {
+		uk_pr_crit("Could not register handler for IPI IRQ %ld\n",
+			   run_irq);
+		return rc;
+	}
+
+	rc = ukplat_irq_register(wakeup_irq, lcpu_ipi_wakeup_handler, NULL);
+	if (unlikely(rc)) {
+		uk_pr_crit("Could not register handler for wakeup IRQ %ld\n",
+			   wakeup_irq);
+		return rc;
+	}
+
+	_lcpu_run_irqv = run_irq;
+	_lcpu_wakeup_irqv = wakeup_irq;
+
+	return 0;
+}
+
+void __weak __noreturn lcpu_entry_default(struct lcpu *this_lcpu)
+{
+	struct lcpu_sargs s_args = this_lcpu->s_args;
+	int rc;
+
+	UK_ASSERT(!lcpu_is_bsp(this_lcpu));
+
+	/* Finish initialization. As there is nothing to return to, we
+	 * just enter halted state if an error occurs.
+	 */
+	rc = lcpu_init(this_lcpu);
+	if (unlikely(rc))
+		lcpu_halt(this_lcpu, rc);
+
+	/* If the user supplied an entry function jump to it */
+	if (s_args.entry && s_args.entry != lcpu_entry_default) {
+		/* Does not return */
+		lcpu_arch_jump_to(s_args.stackp, s_args.entry);
+	} else {
+		/* We are coming from BUSY0 state and want to transition to
+		 * IDLE state. However, there can be functions queued already
+		 * so we have to use a decrement here.
+		 */
+		ukarch_dec(&this_lcpu->state);
+
+		/* Enable IRQs. If there are functions queued we will
+		 * immediately jump to the IRQ handler.
+		 */
+		ukplat_lcpu_enable_irq();
+		while (1) {
+			/* Besides interrupts in general, the halt can be
+			 * interrupted by calls to ukplat_lcpu_run().
+			 */
+			halt();
+		}
+	}
+}
+
+int ukplat_lcpu_start(const __lcpuidx lcpuidx[], unsigned int *num, void *sp[],
+		      const ukplat_lcpu_entry_t entry[], unsigned long flags)
+{
+	__lcpuid this_cpu_id = ukplat_lcpu_id();
+	struct lcpu *lcpu;
+	unsigned int i, n, argi = 0;
+	const int new = LCPU_STATE_INIT;
+	int old;
+	int rc = 0, rc2 = 0;
```

Comment:
This causes a `warning: unused variable`. Perhaps make conditional to `LCPU_ARCH_MULTI_PHASE_STARTUP`?